documentation/man/man3/SQLiteHelper.3
.TH "SQLiteHelper" 3 "Sun Jan 1 2023" "Version 1.0" "File Integrity Checker" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SQLiteHelper \- The SQL Helper class for the SQLite database\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <SQLiteHelper\&.h>\fP
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBAttributeToCheck\fP { \fBHashedVal\fP = 1, \fBFileSize\fP = 2, \fBDateModified\fP = 3, \fBNONE\fP = 4 }"
.br
.RI "Enumerator used to set the file attribute to check for changes or to update the value for\&. "
.ti -1c
.RI "enum \fBTableToCheck\fP { \fBFiles\fP = 1, \fBChanges\fP = 2 }"
.br
.RI "Enumerator used to set the table to check for file attribute changes\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static int \fBcheckIfAttributeIsInChanges\fP (const int file_id, const \fBAttributeToCheck\fP attribute, sqlite3 *\fBdb\fP)"
.br
.RI "Determines if the provided attribute is in the changes table\&. "
.ti -1c
.RI "static int \fBcheckIfAttributeHasChanged\fP (const int file_id, const \fBAttributeToCheck\fP attribute, const \fBTableToCheck\fP table, sqlite3 *\fBdb\fP, const char *char_attr=NULL, const int int_attr=0)"
.br
.RI "Determines if the provided attribute has changed its value\&. "
.ti -1c
.RI "static char * \fBgetCurrentTime\fP ()"
.br
.RI "Will retrieve the current time\&. "
.ti -1c
.RI "static void \fBsetDirectoryName\fP (char *dir_name)"
.br
.RI "Sets the relative directory name\&. "
.ti -1c
.RI "static void \fBsetFileName\fP (char *file_name)"
.br
.RI "Sets the absolute file name\&. "
.ti -1c
.RI "static void \fBsetMachineName\fP (char *machine_name)"
.br
.RI "Sets the machine name where the scanner is running\&. "
.ti -1c
.RI "static int \fBgetDirectoryId\fP (sqlite3 *\fBdb\fP)"
.br
.RI "Retrieves the id of the directory from the directories table\&. "
.ti -1c
.RI "static int \fBgetDirectoryId\fP (sqlite3 *\fBdb\fP, char *dir_name, char *machine_name)"
.br
.RI "Retrieves the id of the directory from the directories table\&. "
.ti -1c
.RI "static int \fBgetFileId\fP (sqlite3 *\fBdb\fP)"
.br
.RI "Returns the id of the file if the file exists in the files table\&. "
.ti -1c
.RI "static int \fBgetFileId\fP (sqlite3 *\fBdb\fP, char *file_name, int dir_id)"
.br
.RI "Returns the id of the file if the file exists in the files table\&. "
.ti -1c
.RI "static int \fBsaveFileInfo\fP (sqlite3 *\fBdb\fP, const int dir_id, const char *file_name, const int file_size=0, const char *date_modified=NULL, const char *hashed_val=NULL, const bool isNew=true, \fBAttributeToCheck\fP attr_to_update=\fBAttributeToCheck::NONE\fP)"
.br
.RI "Entry point to saving the file information\&. "
.in -1c
.SS "Static Private Member Functions"

.in +1c
.ti -1c
.RI "static int \fBdidHashedValueChange\fP (const char *hashed_val, \fBTableToCheck\fP table, const int file_id, sqlite3 *\fBdb\fP)"
.br
.ti -1c
.RI "static int \fBdidFileSizeChange\fP (const int file_size, \fBTableToCheck\fP table, const int file_id, sqlite3 *\fBdb\fP)"
.br
.ti -1c
.RI "static int \fBdidDateModifiedChange\fP (const char *date_modified, \fBTableToCheck\fP table, const int file_id, sqlite3 *\fBdb\fP)"
.br
.ti -1c
.RI "static int \fBsaveNewFileInfo\fP (sqlite3 *\fBdb\fP)"
.br
.ti -1c
.RI "static int \fBsaveExistingFileInfo\fP (sqlite3 *\fBdb\fP, \fBAttributeToCheck\fP attr_to_update, const char *text_val=NULL, const int int_val=0)"
.br
.ti -1c
.RI "static int \fBinsertTextValueToChanges\fP (sqlite3 *\fBdb\fP, const int file_id, const char *text_val, const char *attr_changed)"
.br
.ti -1c
.RI "static int \fBinsertIntValueToChanges\fP (sqlite3 *\fBdb\fP, const int file_id, const int int_val, const char *attr_changed)"
.br
.ti -1c
.RI "static int \fBupdateChangedInFile\fP (sqlite3 *\fBdb\fP, int file_id, \fBAttributeToCheck\fP attr_to_change)"
.br
.in -1c
.SH "Detailed Description"
.PP 
The SQL Helper class for the SQLite database\&. 
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBSQLiteHelper::AttributeToCheck\fP"

.PP
Enumerator used to set the file attribute to check for changes or to update the value for\&. 
.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIHashedVal \fP\fP
.TP
\fB\fIFileSize \fP\fP
.TP
\fB\fIDateModified \fP\fP
.TP
\fB\fINONE \fP\fP
.PP
.nf
26 { HashedVal = 1, FileSize = 2, DateModified = 3, NONE = 4};
.fi
.SS "enum \fBSQLiteHelper::TableToCheck\fP"

.PP
Enumerator used to set the table to check for file attribute changes\&. 
.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIFiles \fP\fP
.TP
\fB\fIChanges \fP\fP
.PP
.nf
31 {Files = 1, Changes = 2};
.fi
.SH "Member Function Documentation"
.PP 
.SS "int SQLiteHelper::checkIfAttributeHasChanged (const int file_id, const \fBAttributeToCheck\fP attribute, const \fBTableToCheck\fP table, sqlite3 * db, const char * char_attr = \fCNULL\fP, const int int_attr = \fC0\fP)\fC [static]\fP"

.PP
Determines if the provided attribute has changed its value\&. The logic in the function will first determine the attribute to check using the attribute parameter and then will call the appropriate function to check if the attribute did change\&. 
.PP
\fBParameters\fP
.RS 4
\fIfile_id\fP The id of the file\&. 
.br
\fIattribute\fP The attribute to check\&. 
.br
\fItable\fP The table to check based on whether the attribute is in the changes or the files table\&. 
.br
\fIdb\fP The sqlite3 database object\&. 
.br
\fIchar_attr\fP The character value attribute to check\&. The default is NULL\&. 
.br
\fIint_attr\fP The integer value attribute to check\&. The default is 0 
.RE
.PP
\fBSee also\fP
.RS 4
\fBAttributeToCheck\fP 
.PP
\fBTableToCheck\fP 
.PP
\fBdidHashedValueChange()\fP 
.PP
\fBdidFileSizeChange()\fP 
.PP
\fBdidDateModifiedChange()\fP 
.RE
.PP
\fBReturns\fP
.RS 4
1 if changed, 0 if not 
.RE
.PP

.PP
.nf
45                                                                                                                                                                                   {
46     int retVal = 0;
47 
48     switch(attribute) {
49     case 1: //Hashedval
50         retVal = didHashedValueChange(char_attr, table, file_id, db);
51         break;
52     case 2: //Filesize
53         retVal = didFileSizeChange(int_attr, table, file_id, db);
54         break;
55     case 3: //Datemodified
56         retVal = didDateModifiedChange(char_attr, table, file_id, db);
57         break;
58     default:
59         retVal = -1;
60     }
61 
62     //fprintf(stderr, "\nReturn value from checkIfAttributeHasChanged is %d", retVal);
63     
64     return retVal;
65 }
.fi
.SS "int SQLiteHelper::checkIfAttributeIsInChanges (const int file_id, const \fBAttributeToCheck\fP attribute, sqlite3 * db)\fC [static]\fP"

.PP
Determines if the provided attribute is in the changes table\&. 
.PP
\fBParameters\fP
.RS 4
\fIfile_id\fP The id of the file\&. 
.br
\fIattribute\fP The attribute to check\&. 
.br
\fIdb\fP The sqlite3 database object\&. 
.RE
.PP
\fBReturns\fP
.RS 4
1 if is in the changes table, 0 if not 
.RE
.PP

.PP
.nf
67                                                                                                               {
68     char* zErrMsg = 0;
69     int rc;
70     sqlite3_stmt* pStmt;
71     char* sql_stmt;
72     
73     int retVal = 0;
74 
75     switch (attribute) {
76     case 1: //Hashedval
77         rc = sqlite3_prepare_v3(db, sql_check_if_hashedval_changed, -1, 0, &pStmt, NULL);
78         break;
79     case 2: //Filesize
80         rc = sqlite3_prepare_v3(db, sql_check_if_filesize_changed, -1, 0, &pStmt, NULL);
81         break;
82     case 3: //Datemodified
83         rc = sqlite3_prepare_v3(db, sql_check_if_datemodified_changed, -1, 0, &pStmt, NULL);
84         break;
85     default:
86         retVal = -1;
87     }
88 
89     if (rc == SQLITE_OK) {
90         sqlite3_bind_int(pStmt, 1, file_id);
91     }
92     else {
93 
94         fprintf(stderr, "checkIfAttributeIsInChanges::Failed to execute statement: %s\n", sqlite3_errmsg(db));
95         return -1;
96     }
97 
98     while (sqlite3_step(pStmt) == SQLITE_ROW)
99     {
100         retVal = sqlite3_column_int(pStmt, 0);
101     }
102 
103     return retVal;
104 }
.fi
.SS "int SQLiteHelper::didDateModifiedChange (const char * date_modified, \fBTableToCheck\fP table, const int file_id, sqlite3 * db)\fC [static]\fP, \fC [private]\fP"

.PP
\fBParameters\fP
.RS 4
\fIdate_modified\fP 
.br
\fItable\fP 
.br
\fIfile_id\fP 
.br
\fIdb\fP 
.RE
.PP
\fBReturns\fP
.RS 4
.RE
.PP

.PP
.nf
189                                                                                                                      {
190     int retVal = 0;
191 
192     char* zErrMsg = 0;
193     int rc;
194     sqlite3_stmt* pStmt;
195    
196 
197     if (table == 1) {
198         rc = sqlite3_prepare_v3(db, sql_check_if_datemodified_changed_in_files, -1, 0, &pStmt, NULL);
199     }
200     else if (table == 2) {
201         rc = sqlite3_prepare_v3(db, sql_check_if_value_changed_in_changes_textval, -1, 0, &pStmt, NULL);
202     }
203     else
204     {
205         return -1;
206     }
207 
208 
209     if (rc == SQLITE_OK) {
210         sqlite3_bind_text(pStmt, 1, date_modified, -1, NULL);
211         sqlite3_bind_int(pStmt, 2, file_id);
212     }
213     else {
214 
215         fprintf(stderr, "didDateModifiedChange::Failed to execute statement: %s\n", sqlite3_errmsg(db));
216         return -1;
217     }
218 
219     while (sqlite3_step(pStmt) == SQLITE_ROW)
220     {
221         retVal = sqlite3_column_int(pStmt, 0);
222     }
223 
224     sqlite3_finalize(pStmt);
225 
226     return retVal;
227 }
.fi
.SS "int SQLiteHelper::didFileSizeChange (const int file_size, \fBTableToCheck\fP table, const int file_id, sqlite3 * db)\fC [static]\fP, \fC [private]\fP"

.PP
\fBParameters\fP
.RS 4
\fIfile_size\fP 
.br
\fItable\fP 
.br
\fIfile_id\fP 
.br
\fIdb\fP 
.RE
.PP
\fBReturns\fP
.RS 4
.RE
.PP

.PP
.nf
149                                                                                                            {
150     int retVal = 0;
151 
152     char* zErrMsg = 0;
153     int rc;
154     sqlite3_stmt* pStmt;
155     
156 
157     if (table == 1) {
158         rc = sqlite3_prepare_v3(db, sql_check_if_filesize_changed_in_files, -1, 0, &pStmt, NULL);
159     }
160     else if (table == 2) {
161         rc = sqlite3_prepare_v3(db, sql_check_if_value_changed_in_changes_intval, -1, 0, &pStmt, NULL);
162     }
163     else
164     {
165         return -1;
166     }
167 
168 
169     if (rc == SQLITE_OK) {
170         sqlite3_bind_int(pStmt, 1, file_size);
171         sqlite3_bind_int(pStmt, 2, file_id);
172     }
173     else {
174 

src/SQLiteHelper.cpp
/**
* @file SQLiteHelper.cpp
*/

#include "../headers/SQLiteHelper.h"

using namespace::std;

char* sql_check_if_filesize_changed = "select filesizechanged from files where id=?; ";
char* sql_check_if_filesize_changed_in_files = "select id from files where filesize = ? and id = ?; ";

char* sql_check_if_hashedval_changed = "select hashedvaluechanged from files where id=?; ";
char* sql_check_if_hashedval_changed_in_files = "select id from files where hashedvalue = ? and id = ?; ";

char* sql_check_if_datemodified_changed = "select lastmodifiedchanged from files where id=?; ";
char* sql_check_if_datemodified_changed_in_files = "select id from files where lastmodified = ? and id = ?; ";

char* sql_check_if_value_changed_in_changes_textval = "select id from changes where fileid = ? and textvalue = ?; ";
char* sql_check_if_value_changed_in_changes_intval = "select id from changes where fileid = ? and intvalue = ?; ";

char* sql_get_file_id = "select id from files where filename=? and dirid=?; ";
char* sql_get_directory_id = "select id from directories where dirname=? and machinename= ?; ";

char* sql_save_newfile_nohashedval = "insert into files (dirid, dateadded, filename, filesize, lastmodified) VALUES (?,?,?,?,?); ";
char* sql_save_newfile_hashedval = "insert into files (dirid, dateadded, filename, filesize, lastmodified, hashval) VALUES (?,?,?,?,?,?); ";

char* sql_save_textval_in_changes = "INSERT INTO changes (fileid,datechanged,textvalue,attributechanged) VALUES (?,?,?,?); ";
char* sql_save_intval_in_changes = "INSERT INTO changes (fileid,datechanged,intvalue,attributechanged) VALUES (?,?,?,?); ";

char* sql_set_inchanges_hashedval = "UPDATE files SET hashedvaluechanged=1 WHERE id=?; ";
char* sql_set_inchanges_filesize = "UPDATE files SET filesizechanged=1 WHERE id=?; ";
char* sql_set_inchanges_lastmodified = "UPDATE files SET lastmodifiedchanged=1 WHERE id=?; ";

char* _dirname;
char* _filename;
char* _machinename;
char* _datemodified;
char* _hashedval;
int _filesize;
int _dirid;

sqlite3_stmt* pStmt;


int SQLiteHelper::checkIfAttributeHasChanged(const int file_id, const AttributeToCheck attribute, const TableToCheck table, sqlite3* db, const char* char_attr, const int int_attr) {
    int retVal = 0;

    switch(attribute) {
    case 1: //Hashedval
        retVal = didHashedValueChange(char_attr, table, file_id, db);
        break;
    case 2: //Filesize
        retVal = didFileSizeChange(int_attr, table, file_id, db);
        break;
    case 3: //Datemodified
        retVal = didDateModifiedChange(char_attr, table, file_id, db);
        break;
    default:
        retVal = -1;
    }

    //fprintf(stderr, "\nReturn value from checkIfAttributeHasChanged is %d", retVal);
    
    return retVal;
}

int SQLiteHelper::checkIfAttributeIsInChanges(const int file_id, const AttributeToCheck attribute, sqlite3* db) {
    char* zErrMsg = 0;
    int rc;
    sqlite3_stmt* pStmt;
    char* sql_stmt;
    
    int retVal = 0;

    switch (attribute) {
    case 1: //Hashedval
        rc = sqlite3_prepare_v3(db, sql_check_if_hashedval_changed, -1, 0, &pStmt, NULL);
        break;
    case 2: //Filesize
        rc = sqlite3_prepare_v3(db, sql_check_if_filesize_changed, -1, 0, &pStmt, NULL);
        break;
    case 3: //Datemodified
        rc = sqlite3_prepare_v3(db, sql_check_if_datemodified_changed, -1, 0, &pStmt, NULL);
        break;
    default:
        retVal = -1;
    }

    if (rc == SQLITE_OK) {
        sqlite3_bind_int(pStmt, 1, file_id);
    }
    else {

        fprintf(stderr, "checkIfAttributeIsInChanges::Failed to execute statement: %s\n", sqlite3_errmsg(db));
        return -1;
    }

    while (sqlite3_step(pStmt) == SQLITE_ROW)
    {
        retVal = sqlite3_column_int(pStmt, 0);
    }

    return retVal;
}

int SQLiteHelper::didHashedValueChange(const char* hashed_val, TableToCheck table, const int file_id, sqlite3* db) {
    int retVal = 0;

    char* zErrMsg = 0;
    int rc;
    sqlite3_stmt* pStmt;
    

    if (table == 1) {
        rc = sqlite3_prepare_v3(db, sql_check_if_hashedval_changed_in_files, -1, 0, &pStmt, NULL);
    }
    else if (table == 2) {
        rc = sqlite3_prepare_v3(db, sql_check_if_value_changed_in_changes_textval, -1, 0, &pStmt, NULL);
    }
    else
    {
        return -1;
    }
    

    if (rc == SQLITE_OK) {
        
        sqlite3_bind_text(pStmt, 1, hashed_val, -1, NULL);
        sqlite3_bind_int(pStmt, 2, file_id);
        
    }
    else {

        fprintf(stderr, "didHashedValueChange::Failed to execute statement: %s\n", sqlite3_errmsg(db));
        return -1;
    }

    while (sqlite3_step(pStmt) == SQLITE_ROW)
    {
        retVal = sqlite3_column_int(pStmt, 0);
    }

    sqlite3_finalize(pStmt);

    
    return retVal;
}

int SQLiteHelper::didFileSizeChange(const int file_size, TableToCheck table, const int file_id, sqlite3* db) {
    int retVal = 0;

    char* zErrMsg = 0;
    int rc;
    sqlite3_stmt* pStmt;
    

    if (table == 1) {
        rc = sqlite3_prepare_v3(db, sql_check_if_filesize_changed_in_files, -1, 0, &pStmt, NULL);
    }
    else if (table == 2) {
        rc = sqlite3_prepare_v3(db, sql_check_if_value_changed_in_changes_intval, -1, 0, &pStmt, NULL);
    }
    else
    {
        return -1;
    }


    if (rc == SQLITE_OK) {
        sqlite3_bind_int(pStmt, 1, file_size);
        sqlite3_bind_int(pStmt, 2, file_id);
    }
    else {

        fprintf(stderr, "didFileSizeChange::Failed to execute statement: %s\n", sqlite3_errmsg(db));
        return -1;
    }

    while (sqlite3_step(pStmt) == SQLITE_ROW)
    {
        retVal = sqlite3_column_int(pStmt, 0);
    }

    sqlite3_finalize(pStmt);

    return retVal;
}

int SQLiteHelper::didDateModifiedChange(const char* date_modified, TableToCheck table, const int file_id, sqlite3* db) {
    int retVal = 0;

    char* zErrMsg = 0;
    int rc;
    sqlite3_stmt* pStmt;
   

    if (table == 1) {
        rc = sqlite3_prepare_v3(db, sql_check_if_datemodified_changed_in_files, -1, 0, &pStmt, NULL);
    }
    else if (table == 2) {
        rc = sqlite3_prepare_v3(db, sql_check_if_value_changed_in_changes_textval, -1, 0, &pStmt, NULL);
    }
    else
    {
        return -1;
    }


    if (rc == SQLITE_OK) {
        sqlite3_bind_text(pStmt, 1, date_modified, -1, NULL);
        sqlite3_bind_int(pStmt, 2, file_id);
    }
    else {

        fprintf(stderr, "didDateModifiedChange::Failed to execute statement: %s\n", sqlite3_errmsg(db));
        return -1;
    }

    while (sqlite3_step(pStmt) == SQLITE_ROW)
    {
        retVal = sqlite3_column_int(pStmt, 0);
    }

    sqlite3_finalize(pStmt);

    return retVal;
}


void SQLiteHelper::setDirectoryName(char* dir_name) {
    _dirname = (char*)malloc(strlen(dir_name) + 1);
    strcpy(_dirname, dir_name);
}

void SQLiteHelper::setFileName(char* file_name) {
    _filename = (char*)malloc(strlen(file_name) + 1);
    strcpy(_filename, file_name);
}

void SQLiteHelper::setMachineName(char* machine_name) {
    _machinename = (char*)malloc(strlen(machine_name) + 1);
    strcpy(_machinename, machine_name);
}

int SQLiteHelper::getDirectoryId(sqlite3* db) {
    char* zErrMsg = 0;
    int rc;
    sqlite3_stmt* pStmt;
    int retVal = 0;

    rc = sqlite3_prepare_v3(db, sql_get_directory_id, -1, 0, &pStmt, NULL);

    if (rc == SQLITE_OK) {
        sqlite3_bind_text(pStmt, 1, _dirname, -1, NULL);
        sqlite3_bind_text(pStmt, 2, _machinename, -1, NULL);
    }
    else {

        fprintf(stderr, "getDirectoryId::Failed to execute statement: %s\n", sqlite3_errmsg(db));
        return -1;
    }

    while (sqlite3_step(pStmt) == SQLITE_ROW)
    {
        retVal = sqlite3_column_int(pStmt, 0);
    }

    sqlite3_finalize(pStmt);

    return retVal;
}

int SQLiteHelper::getDirectoryId(sqlite3* db, char* dir_name, char* machine_name) {
    int retVal = 0;

    _dirname = (char*)malloc(strlen(dir_name) + 1);
    strcpy(_dirname, dir_name);

    _machinename = (char*)malloc(strlen(machine_name) + 1);
    strcpy(_machinename, machine_name);

    retVal = getDirectoryId(db);

    return retVal;

}

int SQLiteHelper::getFileId(sqlite3* db) {
    char* zErrMsg = 0;
    int rc;
    sqlite3_stmt* pStmt;
    int retVal = 0;

    rc = sqlite3_prepare_v3(db, sql_get_file_id, -1, 0, &pStmt, NULL);

    if (rc == SQLITE_OK) {
        sqlite3_bind_text(pStmt, 1, _filename, -1, NULL);
        sqlite3_bind_int(pStmt, 2, _dirid);
    }
    else {

        fprintf(stderr, "getFileId::Failed to execute statement: %s\n", sqlite3_errmsg(db));
        return -1;
    }

    while (sqlite3_step(pStmt) == SQLITE_ROW)
    {
        retVal = sqlite3_column_int(pStmt, 0);
    }

    sqlite3_finalize(pStmt);

    return retVal;
}

int SQLiteHelper::getFileId(sqlite3* db, char* file_name, int dir_id) {
    int retVal = 0;

    _filename = (char*)malloc(strlen(file_name) + 1);
    strcpy(_filename, file_name);
    _dirid = dir_id;

    retVal = getFileId(db);

    return retVal;
}


/**
 * @brief 
 * 
 * @return char* 
 */
char* SQLiteHelper::getCurrentTime() {
    string retVal;
	char* char_array;

	time_t t = time(0);   // get time now
    tm* now = localtime(&t);
	char date_string[100];
	char time_string[100];

	strftime(date_string, 50, "%B %d, %Y ", now);
	strftime(time_string, 50, "%T", now);

	retVal += date_string;
	retVal += time_string;

	int n = retVal.length();
	char_array = new char[n + 1];

	strcpy(char_array, retVal.c_str());
	
	return char_array;
}

int SQLiteHelper::saveFileInfo(sqlite3* db, const int dir_id, const char* file_name, const int file_size, const char* date_modified, const char* hashed_val, const bool isNew, AttributeToCheck attr_to_update) {
    int retVal = 0;
    int hasChanged = 0;

    _dirid = dir_id;
    _filename = (char*)malloc(strlen(file_name)+1);
    strcpy(_filename,file_name);
    _filesize = file_size;

    if (date_modified != NULL) {
        _datemodified = (char*)malloc(strlen(date_modified) + 1);
        strcpy(_datemodified, date_modified);
    }
    
    if (hashed_val != NULL) {
        //fprintf(stderr, "\nGetting ready to store hashedvalue into _hashedval with value of %s\n", _hashedval);
        _hashedval = (char*)malloc(strlen(hashed_val) + 1);
        strcpy(_hashedval, (char*)hashed_val);
    }
    
    if (isNew) {
        saveNewFileInfo(db);
    }
    else {
        switch (attr_to_update)
        {
        case 1:
            //fprintf(stderr, "\nGetting ready to save hashedvalue of %s\n", _hashedval);
            saveExistingFileInfo(db, attr_to_update,_hashedval);
            break;
        case 2:
            saveExistingFileInfo(db, attr_to_update,NULL,_filesize);
            break;
        case 3:
            saveExistingFileInfo(db, attr_to_update, _datemodified);
            break;
        default:
            break;
        }
        //saveExistingFileInfo(db, attr_to_update);
    }

    return retVal;
}

int SQLiteHelper::saveNewFileInfo(sqlite3* db) {
    int retVal = 0;
    char* zErrMsg = 0;
    int rc;
    sqlite3_stmt* pStmt;
    
    char* current_time = getCurrentTime();

    if (_hashedval == NULL) {
        rc = sqlite3_prepare_v3(db, sql_save_newfile_nohashedval, -1, 0, &pStmt, NULL);
    }
    else {
        rc = sqlite3_prepare_v3(db, sql_save_newfile_hashedval, -1, 0, &pStmt, NULL);
    }

    //dirid, dateadded, filename, filesize, lastmodified
    if (rc == SQLITE_OK) {
        sqlite3_bind_int(pStmt, 1, _dirid);
        sqlite3_bind_text(pStmt, 2, current_time, -1, NULL);
        sqlite3_bind_text(pStmt, 3, _filename, -1, NULL);
        sqlite3_bind_int(pStmt, 4, _filesize);
        sqlite3_bind_text(pStmt, 5, _datemodified, -1, NULL);
        
        if (_hashedval != NULL) {
            sqlite3_bind_text(pStmt, 6, _hashedval, -1, NULL);
        }
    }
    else {

        //fprintf(stderr, "saveNewFileInfo::Failed to execute statement: %s\n", sqlite3_errmsg(db));
        retVal = -1;
    }

    rc = sqlite3_step(pStmt);

    if (rc == SQLITE_BUSY) {
        cout << "\nDatabase is busy" << endl;
        while (rc == SQLITE_BUSY)
        {
            rc = sqlite3_step(pStmt);
        }
    }
    else if (rc == SQLITE_ERROR) {
        printf("\nIn SQLiteHelper::saveNewFileInfo execution failed: %s\n", sqlite3_errmsg(db));
    }
    else if (rc == SQLITE_MISUSE) {
        cout << "\nA MISUSE has been deteced in SQLiteHelper::saveNewFileInfo." << endl;
    }
    else if (rc == SQLITE_DONE) {
        cout << "\nsqlite3_step executed successfully in SQLiteHelper::saveNewFileInfo." << endl;
    }
    else {
        cout << "\nNot sure what happened in SQLiteHelper::saveNewFileInfo." << endl;
    }

    sqlite3_finalize(pStmt);
    
    return retVal;
}

int SQLiteHelper::saveExistingFileInfo(sqlite3* db, AttributeToCheck attr_to_update, const char* text_val, const int int_val) {
    int retVal = 0;

    char* zErrMsg = 0;
    int rc;
    sqlite3_stmt* pStmt;
    
    char* current_time = getCurrentTime();
    int fileId = getFileId(db);
    
    int in_changes = checkIfAttributeIsInChanges(fileId, attr_to_update, db);
    int has_changes = 0;

    switch (attr_to_update)
    {
    case 1:

        if (!in_changes) {
            has_changes = checkIfAttributeHasChanged(fileId, attr_to_update, TableToCheck::Files, db, text_val);
        }
        else {
            has_changes = checkIfAttributeHasChanged(fileId, attr_to_update, TableToCheck::Changes, db, text_val);
        }
        
        if (has_changes == 0) {
            //fprintf(stderr, "\nInside has_changes with value of %d and text value %s\n", has_changes, text_val);
            retVal = insertTextValueToChanges(db, fileId, text_val,"hashedvalue");
        }
        break;
    case 2:
        
        if (!in_changes) {
            has_changes = checkIfAttributeHasChanged(fileId, attr_to_update, TableToCheck::Files, db, NULL, int_val);
        }
        else {
            has_changes = checkIfAttributeHasChanged(fileId, attr_to_update, TableToCheck::Changes, db, NULL, int_val);
        }
        
        if (has_changes == 0) {
            //fprintf(stderr, "\nInside has_changes with value of %d and int value %d\n", has_changes, int_val);
            retVal = insertIntValueToChanges(db, fileId, int_val, "filesize");
        }
        break;
    case 3:

        if (!in_changes) {
            has_changes = checkIfAttributeHasChanged(fileId, attr_to_update, TableToCheck::Files, db, text_val);
        }
        else {
            has_changes = checkIfAttributeHasChanged(fileId, attr_to_update, TableToCheck::Changes, db, text_val);
        }
        
        if (has_changes == 0) {
            //fprintf(stderr, "\nInside has_changes with value of %d and text value %s\n", has_changes, text_val);
            retVal = insertTextValueToChanges(db, fileId, text_val,"lastmodified");
        }
        break;
    default:
        break;
    }

    if (!in_changes && has_changes == 0) {
        retVal = updateChangedInFile(db, fileId, attr_to_update);
    }

    return retVal;
}

int SQLiteHelper::insertTextValueToChanges(sqlite3* db, const int file_id, const char* text_val, const char* attr_changed) {
    char* zErrMsg = 0;
    int rc;
    sqlite3_stmt* pStmt;
    char* current_time = getCurrentTime();
    
    rc = sqlite3_prepare_v3(db, sql_save_textval_in_changes, -1, 0, &pStmt, NULL);

    if (rc == SQLITE_OK) {
        sqlite3_bind_int(pStmt, 1, file_id);
        sqlite3_bind_text(pStmt, 2, current_time, -1, NULL);
        sqlite3_bind_text(pStmt, 3, text_val, -1, NULL);
        sqlite3_bind_text(pStmt, 4, attr_changed, -1, NULL);
    }
    else {

        fprintf(stderr, "insertTextValueToChanges::Failed to execute statement: %s\n", sqlite3_errmsg(db));
        sqlite3_finalize(pStmt);
        return -1;
    }

    rc = sqlite3_step(pStmt);

    if (rc == SQLITE_BUSY) {
        cout << "\nDatabase is busy" << endl;
        while (rc == SQLITE_BUSY)
        {
            rc = sqlite3_step(pStmt);
        }
    }
    else if (rc == SQLITE_ERROR) {
        printf("\nIn insertTextValuetoChanges execution failed: %s\n", sqlite3_errmsg(db));
    }
    else if (rc == SQLITE_MISUSE) {
        cout << "\nA MISUSE has been deteced in insertTextValuetoChanges." << endl;
    }
    else if (rc == SQLITE_DONE) {
        cout << "\nsqlite3_step executed successfully." << endl;
    }
    else {
        cout << "\nNot sure what happened in insertTextValuetoChanges." << endl;
    }

    sqlite3_finalize(pStmt);

    return 0;
}

int SQLiteHelper::insertIntValueToChanges(sqlite3* db, const int file_id, const int int_val, const char* attr_changed) {
    char* zErrMsg = 0;
    int rc;
    sqlite3_stmt* pStmt;
    char* current_time = getCurrentTime();

    rc = sqlite3_prepare_v3(db, sql_save_intval_in_changes, -1, 0, &pStmt, NULL);

    if (rc == SQLITE_OK) {
        sqlite3_bind_int(pStmt, 1, file_id);
        sqlite3_bind_text(pStmt, 2, current_time, -1, NULL);
        sqlite3_bind_int(pStmt, 3, int_val);
        sqlite3_bind_text(pStmt, 4, attr_changed, -1, NULL);
    }
    else {

        fprintf(stderr, "insertIntValueToChanges::Failed to execute statement: %s\n", sqlite3_errmsg(db));
        sqlite3_finalize(pStmt);
        return -1;
    }

    rc = sqlite3_step(pStmt);

    if (rc == SQLITE_BUSY) {
        cout << "\nDatabase is busy" << endl;
        while (rc == SQLITE_BUSY)
        {
            rc = sqlite3_step(pStmt);
        }
    }
    else if (rc == SQLITE_ERROR) {
        printf("\nIn insertIntValueToChanges execution failed: %s\n", sqlite3_errmsg(db));
    }
    else if (rc == SQLITE_MISUSE) {
        cout << "\nA MISUSE has been deteced in insertIntValueToChanges." << endl;
    }
    else if (rc == SQLITE_DONE) {
        cout << "\nsqlite3_step executed successfully." << endl;
    }
    else {
        cout << "\nNot sure what happened in insertIntValueToChanges." << endl;
    }

    sqlite3_finalize(pStmt);

    return 0;
}

int SQLiteHelper::updateChangedInFile(sqlite3* db, int file_id, AttributeToCheck attr_to_change) {
    char* zErrMsg = 0;
    int rc;
    sqlite3_stmt* pStmt;
    int retVal = 0;
    char* current_time = getCurrentTime();
    char* sql_stmt;
    
    switch (attr_to_change)
    {
    case 1:
        rc = sqlite3_prepare_v3(db, sql_set_inchanges_hashedval, -1, 0, &pStmt, NULL);
        break;
    case 2:
        rc = sqlite3_prepare_v3(db, sql_set_inchanges_filesize, -1, 0, &pStmt, NULL);
        break;
    case 3:
        rc = sqlite3_prepare_v3(db, sql_set_inchanges_lastmodified, -1, 0, &pStmt, NULL);
        break;
    default:
        return -1;
        cout << "Unknown attribute. Nothing saved." << endl;
        break;
    }

    if (rc == SQLITE_OK) {
        sqlite3_bind_int(pStmt, 1, file_id);
    }
    else {
        fprintf(stderr, "updateChangedInFiles:Failed to execute statement: %s\n", sqlite3_errmsg(db));
        sqlite3_finalize(pStmt);
        return -1;
    }

    rc = sqlite3_step(pStmt);

    if (rc == SQLITE_BUSY) {
        cout << "\nDatabase is busy" << endl;
        while (rc == SQLITE_BUSY)
        {
            rc = sqlite3_step(pStmt);
        }
    }
    else if (rc == SQLITE_ERROR) {
        printf("\nIn updateChangedInFiles execution failed: %s\n", sqlite3_errmsg(db));
    }
    else if (rc == SQLITE_MISUSE) {
        cout << "\nA MISUSE has been deteced in updateChangedInFiles." << endl;
    }
    else if (rc == SQLITE_DONE) {
        cout << "\nsqlite3_step executed successfully." << endl;
    }
    else {
        cout << "\nNot sure what happened in updateChangedInFiles." << endl;
    }

    sqlite3_finalize(pStmt);

    return 0;
}